src/mnesia_api.erl

```diff
@@ -72,47 +74,46 @@ restore_mnesia(Path) ->
             {ok, ""};
         {aborted, {no_exists, Table}} ->
             String = io_lib:format("~sTable ~p does not exist.", [ErrorString, Table]),
-            {error, {table_does_not_exist, String}};
+            {table_does_not_exist, String};
         {aborted, enoent} ->
             String = ErrorString ++ "File not found.",
-            {error, {file_not_found, String}};
+            {file_not_found, String};
         {aborted, {not_a_log_file, Filename}} ->
             String = "Wrong file " ++ Filename ++ " structure",
-            {error, {not_a_log_file_error, String}};
+            {not_a_log_file_error, String};
         {aborted, Reason} ->
             String = io_lib:format("~s~p", [ErrorString, Reason]),
-            {error, {cannot_restore, String}}
+            {cannot_restore, String}
     end.
 
--spec load_mnesia(file:name()) ->
-    {error, {cannot_load | bad_file_format | file_not_found, io_lib:chars()}} | {ok, []}.
+-spec load_mnesia(file:name()) -> {load_error(), io_lib:chars()} | {ok, []}.
 load_mnesia(Path) ->
     case mnesia:load_textfile(Path) of
         {atomic, ok} ->
             {ok, ""};
         {error, bad_header} ->
-            {error, {bad_file_format, "File has wrong format"}};
+            {bad_file_format, "File has wrong format"};
         {error, read} ->
-            {error, {bad_file_format, "File has wrong format"}};
+            {bad_file_format, "File has wrong format"};
         {error, open} ->
-            {error, {file_not_found, "File was not found"}};
+            {file_not_found, "File was not found"};
         {error, Reason} ->
             String = io_lib:format("Can't load dump in ~p at node ~p: ~p",
                                    [filename:absname(Path), node(), Reason]),
-            {error, {cannot_load, String}}
+            {cannot_load, String}
     end.
 
--spec mnesia_change_nodename(string(), string(), _, _) -> {ok, _} | {error, _}.
-mnesia_change_nodename(FromString, ToString, Source, Target) ->
-    From = list_to_atom(FromString),
-    To = list_to_atom(ToString),
+-spec mnesia_change_nodename(atom() | string(), atom() | string(), _, _) ->
+                                {ok, _} | {change_error(), io_lib:chars()}.
+mnesia_change_nodename(From, To, Source, Target) when is_list(From), is_list(To) ->
+    mnesia_change_nodename(list_to_atom(From), list_to_atom(To), Source, Target);
```

Comment:
This clause might be unreachable, because we always convert the node names to atoms, right?